### PR TITLE
Add support for creating certificates from CSRs

### PR DESCRIPTION
**Pull request by djc**
Depends on https://github.com/rusticata/x509-parser/pull/58 and (transitively) https://github.com/rusticata/oid-registry/pull/1, but if you have any early feedback that would be great!

### CODE REVIEW

**Comment by djc** [comment]
The upstream PRs have been merged (but not yet released).

**Comment by est31** [comment]
Thanks for this PR! I'll review it more closely later but as a remark for the start, I'm not sure if the `from_name` function and the TryFrom impl should be added. This makes the public rcgen API use types from the `x509-parser` crate which means that semver breaking updates cause us to have them too. In general though, it's good to have CSR support in rcgen.

**Comment by est31** [comment]
Also: could you add tests? :)

**Comment by djc** [comment]
About the semver compatibility issues:

- `DistinguishedName::from_name()` is actually private, so I don't think that should be a concern?
- The main entry point `CertificateSigningRequest::from_x509()` itself exposes an x509-parser type in the public API. If it would take DER or PEM bytes, that means either (a) rcgen should do signature verification or (b) the CSR has to be parsed twice if the users wants to also do signature verification. What do you prefer, or do you see another option?

Will add some tests.

**Comment by est31** [comment]
> from_name() is actually private

good point, case rested.

> either (a) rcgen should do signature verification or (b) the CSR has to be parsed twice if the users wants to also do signature verification. What do you prefer, or do you see another option?

Is it hard to do (a)? rcgen's main task is to create certificates, where you can't do much wrong, but doing signature verification wrong doesn't seem *that* risky, as an attacker having a certificate for a public key they don't own doesn't give them *that* many more capabilities. But the same time, it's not the goal of rcgen to verify things, but to create them.

**Comment by djc** [comment]
I changed it to verify the signature using the signature verification feature from x509-parser (which I wrote for their CSR support -- this seemed more convenient than reproducing the verification logic in rcgen. I've also moved the `TryFrom` impl to a private `try_from()` method on `SanType`. Let me know if this is okay, then I'll write up some tests next.

**Comment by est31** [comment]
@djc sounds like a plan! :+1:

**Comment by djc** [comment]
Updated to use x509-parser 0.9.

**Comment by est31** [comment]
Wonderful! Will take a look at it later today.

File: src/lib.rs [hunk 1]
```diff
@@ -156,6 +156,19 @@ pub enum SanType {
 }
 
 impl SanType {
+	#[cfg(feature = "x509-parser")]
+	fn try_from(name: &x509_parser::extensions::GeneralName<'_>) -> Result<Self, RcgenError> {
```
Comment (est31) [inline src/lib.rs:160]: Could you rename this to `try_from_general` or something? `try_from` is very generic, makes mainly sense if part of the TryFrom trait impl. Also: colon placement.

File: src/lib.rs [hunk 2]
```diff
@@ -403,6 +416,31 @@ impl DistinguishedName {
 			iter :self.order.iter()
 		}
 	}
+
+	#[cfg(feature = "x509-parser")]
+	fn from_name(name: &x509_parser::x509::X509Name) -> Result<Self, RcgenError> {
```
Comment (est31) [inline src/lib.rs:421]: colon placement

File: src/lib.rs [hunk 3]
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
...
+
+/// Data for a certificate signing request
+#[allow(missing_docs)]
+pub struct CertificateSigningRequest {
+	pub params :CertificateParams,
+	pub public_key :PublicKey,
+}
+
+impl CertificateSigningRequest {
+	/// Parse certificate signing request from the ASCII PEM format
```
Comment (est31) [inline src/lib.rs:490]:
```suggestion
	/// Parse a certificate signing request from the ASCII PEM format
```

File: src/lib.rs [hunk 4]
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
...
+	///
+	/// See `from_der()` for more details.
+	#[cfg(all(feature = "pem", feature = "x509-parser"))]
+	pub fn from_pem(pem_str :&str) -> Result<Self, RcgenError> {
+		let csr = pem::parse(pem_str)
+			.or(Err(RcgenError::CouldNotParseCertificationRequest))?;
+		Self::from_der(&csr.contents)
+	}
+
+	/// Parse certificate signing request from DER-encoded bytes
```
Comment (est31) [inline src/lib.rs:500]:
```suggestion
	/// Parse a certificate signing request from DER-encoded bytes
```

File: src/lib.rs [hunk 5]
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
...
+	///
+	/// Currently, this only supports the `Subject Alternative Name` extension.
```
Comment (est31) [inline src/lib.rs:502]: Maybe say here that for other extensions, an error is returned.

File: src/lib.rs [hunk 6]
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
...
+			public_key: PublicKey { alg, raw },
+		})
+	}
+	/// Serializes the requested certificate, signed with another certificate's key, in binary DER format
+	pub fn serialize_der_with_signer(&self, ca :&Certificate) -> Result<Vec<u8>, RcgenError> {
+		self.params.serialize_der_with_signer(&self.public_key, ca)
+	}
+	/// Serializes the requested certificate, signed with another certificate's key, to the ASCII PEM format
+	///
+	/// *This function is only available if rcgen is built with the "pem" feature*
```
Comment (est31) [inline src/lib.rs:549]: Thanks for adding this! :heart:

**Review by est31 — CHANGES_REQUESTED** [review]
Ok except for some nits.

**Comment by djc** [comment]
Addressed all the nits.

**Review by est31 — APPROVED** [review]
Thanks! It's a bit sad that the nom update brought bit-vec into our lockfile which has way too many dependencies for the simple task it's doing, but that's more of a long term problem. Merging.

**Comment by djc** [comment]
Yeah, x509-parser should maybe default to depending on nom with default-features = false... Thanks for merging! Do you plan on publishing a new release soon?

**Comment by est31** [comment]
@djc waiting for travis CI to pass, then I'll release. After that I'll switch to github actions because I don't have patience :).

**Comment by djc** [comment]
Yeah, I was surprised you don't have Actions here, yet!

**Comment by est31** [comment]
0.8.6 is out!

**Comment by djc** [comment]
Thanks!

**Pull request merged**